File: Java_Kalk/src/tree/Binarytreebasic.java
package tree;
import dati.*;
import java.util.ArrayList;
import java.lang.IllegalArgumentException;

public abstract class Binarytreebasic {
	
	protected static class Nodo{
		protected Tipo info;
		protected Nodo father;
		protected Nodo left;
		protected Nodo right;
		protected Nodo(Tipo t, Nodo f, Nodo l, Nodo r) {
			info=t;
			father=f;
			left=l;
			right=r;
		}
		protected Nodo(Tipo t, Nodo f, Nodo l) {this(t, f, l, null);}
		protected Nodo(Tipo t, Nodo f) {this(t, f, null, null);}
		protected Nodo(Tipo t) {this(t, null, null, null);}
		protected Nodo() {this(null, null, null, null);}
		protected static String stampa(Nodo n) {
			if(n==null)
				return "_";
			return "('"+n.info.toString()+"',"+stampa(n.left)+","+stampa(n.right)+")";
		}
	}
	Nodo root=null;
	private Nodo constrRic(String s, Tipo[] t, Nodo f, int t_i, int s_i) {
		Nodo x=new Nodo(t[t_i], f);
		s_i+=3;
		int s_left=balance_brackets(s, s_i, s_i);
		int s_right=balance_brackets(s, s_i+s_left+2, s_i+s_left+2);
		if(s_left!=0)
			x.left=constrRic(s, t, x, t_i+1, s_i);
		if(s_right!=0) {
			t_i+=n_nodes_stringa(s, s_i)+1;
			x.right=constrRic(s, t, x, t_i, s_i+s_left+2);
		}
		return x;
	}
	public Binarytreebasic(Tipo[] t, String s) throws IllegalArgumentException{
		if(!controlla_input(t, s))
			throw new IllegalArgumentException("input non validi");
		
		root=constrRic(s, t, null, 0, 0);
	}
	public Binarytreebasic() {}
	protected static boolean parser(String s, int begin, int end) {
		if(begin==end)
			return true;
		if(s.charAt(begin)!='(' || s.charAt(++begin)!='*' || s.charAt(++begin)!=',')
			return false;
		if(s.charAt(begin+1)=='_' && s.charAt(begin+2)==',' && s.charAt(begin+3)=='_' && s.charAt(begin+4)==')')
			return true;
		
		begin++;
		boolean a=true, b=true;
		int aux_end;
		
		if(s.charAt(begin)!='_' && s.charAt(begin)!='(')
			return false;
		if(s.charAt(begin)=='(') {
			int l=balance_brackets(s, begin, end);
			if(l==-1)
				return false;
			aux_end=begin+l+1;
			if(s.charAt(aux_end)!=',')
				return false;
			a=parser(s, begin, aux_end);
			begin=aux_end+1;
		}
		else
			begin+=2;
		
		if(s.charAt(begin)!='_' && s.charAt(begin)!='(')
			return false;
		if(s.charAt(begin)=='(') {
			int r=balance_brackets(s, begin, end);
			if(r==-1)
				return false;
			aux_end=begin+r+1;
			if(s.charAt(aux_end)!=')')
				return false;
			b=parser(s, begin, aux_end);
		}
		
		return a && b;
	}
	protected static int balance_brackets(String s, int begin, int end) {
		if(s.charAt(begin)=='_')
			return 0;
		int k=1, m=1;
		if(begin==end)
			end--;
		
		for(; k>0 && begin+m!=end; m++)
			if(s.charAt(begin+m)=='(')
				k++;
			else if(s.charAt(begin+m)==')')
				k--;
		if(k!=0)
			return -1;
		return m-1;
	}
	protected static int n_nodes_stringa(String s, int begin) {
		int count=0, id=1;
		if(s.charAt(begin)=='_')
			return 0;
		for(int i=1; id!=0; i++) {
			if(s.charAt(begin+i)=='*')
				count++;
			else if(s.charAt(begin+i)=='(')
				id++;
			else if(s.charAt(begin+i)==')')
				id--;
		}
		return count;
	}
	protected static Nodo somma(Nodo a, Nodo b, Nodo f) {
		Nodo x=null;
		if(a==null && b!=null) {
			x=new Nodo(b.info, f);
			x.left=somma(a, b.left, x);
			x.right=somma(a, b.right, x);
		}
		else if(a!=null && b==null) {
			x=new Nodo(a.info, f);
			x.left=somma(a.left, b, x);
			x.right=somma(a.right, b, x);
		}
		else if(a!=null && b!=null && a.left==null && a.right==null && b.left==null && b.right==null && a.info.moltiplicazione(-1).equals(b.info))
			return x;
		else if(a!=null && b!=null) {
			x=new Nodo(a.info.somma(b.info), f);
			x.left=somma(a.left, b.left, x);
			x.right=somma(a.right, b.right, x);
		}
		return x;
	}
	protected static Nodo differenza(Nodo a, Nodo b, Nodo f) {
		Nodo x=null;
		if(a==null && b!=null) {
			x=new Nodo(b.info.moltiplicazione(-1), f);
			x.left=differenza(a, b.left, x);
			x.right=differenza(a, b.right, x);
		}
		else if(a!=null && b==null) {
			x=new Nodo(a.info, f);
			x.left=differenza(a.left, b, x);
			x.right=differenza(a.right, b, x);
		}
		else if(a!=null && b!=null && a.left==null && a.right==null && b.left==null && b.right==null && a.info.equals(b.info))
			return x;
		else if(a!=null && b!=null) {
			x=new Nodo(a.info.differenza(b.info), f);
			x.left=differenza(a.left, b.left, x);
			x.right=somma(a.right, b.right, x);
		}
		return x;
	}
	protected static Nodo moltiplicazione(Nodo a, int k, Nodo f) {
		Nodo x=null;
		if(a!=null) {
			if(k==0)
				return new Nodo(a.info.moltiplicazione(0));
			x=new Nodo(a.info.moltiplicazione(k), f);
			x.left=moltiplicazione(a.left, k, x);
			x.right=moltiplicazione(a.right, k, x);
		}
		return x;
	}
	protected static Nodo divisione(Nodo a, int k, Nodo f) {
		Nodo x=null;
		if(a!=null) {
			x=new Nodo(a.info.divisione(k), f);
			x.left=divisione(a.left, k, x);
			x.right=divisione(a.right, k, x);
		}
		return x;
	}
	protected static ArrayList<Tipo> nodes(Nodo n) {
		ArrayList<Tipo> a=new ArrayList<Tipo>(), b=new ArrayList<Tipo>();
		if(n==null)
			return a;
		if(n.left==null && n.right==null) {
			a.add(n.info);
			return a;
		}
		ArrayList<Tipo> c=new ArrayList<Tipo>();
		c.add(n.info);
		a=nodes(n.left);
		b=nodes(n.right);
		a.addAll(b);
		c.addAll(a);
		return c;
	}
	protected static boolean controlla_percorso(String s) {
		for(int i=0; i<s.length(); i++)
			if(s.charAt(i)!='0' && s.charAt(i)!='1')
				return false;
		return true;
	}
	protected static boolean controlla_input(Tipo[] t, String s) {
		for(int i=0; i<t.length; i++)
			if(t[0].getClass()!=t[i].getClass())
				return false;
		return parser(s, 0, s.length()) && (t.length==n_nodes_stringa(s, 0)) && (s.length()==0 || s.length()==5 || s.length()%6==1);
	}
	protected static String tree_to_string(Nodo r) {
		if(r==null)
			return "_";
		return "(*,"+tree_to_string(r.left)+","+tree_to_string(r.right)+")";
	}
	public static int n_nodes(Nodo n) {
		if(n==null)
			return 0;
		return 1+n_nodes(n.left)+n_nodes(n.right);
	}
	public Tipo cerca(String s) throws IllegalArgumentException{
		Nodo x=root;
		if(!controlla_percorso(s) || x==null)
			throw new IllegalArgumentException("percorso non valido");
		
		for(int i=0; i<s.length(); i++)
			if(s.charAt(i)=='0')
				x=x.left;
			else
				x=x.right;
		if(x!=null)
			return x.info;
		System.out.println("Elemento non trovato\n");
		return null;
	}
	public String struttura_tree() {
		return tree_to_string(root);
	}
	public String toString() {
		if(root!=null)
			return Nodo.stampa(root);
		return new String();
	}
	
	public abstract void add(Tipo t);
	public abstract Tipo remove(String s);
	public abstract Tipo remove(Tipo t);
	public abstract Tipo search(Tipo t);
	public abstract Binarytreebasic somma(Binarytreebasic b);
	public abstract Binarytreebasic differenza(Binarytreebasic b);
	public abstract Binarytreebasic moltiplicazione(int p);
	public abstract Binarytreebasic divisione(int p);
}

File: binarytreebasic.cpp
#include "binarytreebasic.h"

string binarytreebasic::nodo::stampa(binarytreebasic::nodo* n){
    if(!(n))
        return "_";
    return "('"+n->info->to_stringa()+"',"+stampa(n->left)+","+stampa(n->right)+")";
}
bool binarytreebasic::parser(string::iterator begin, string::iterator end){
    if(begin==end)
        return true;
    if(*begin!='(' || *++begin!='*' || *++begin!=',')
        return false;
    if(*(begin+1)=='_' && *(begin+2)==',' && *(begin+3)=='_' && *(begin+4)==')')
        return true;

    begin++;
    bool a=true, b=true;
    string::iterator aux_end;

    if(*(begin)!='_' && *(begin)!='(')
        return false;
    if(*(begin)=='('){
        int l=balance_brackets(begin, end);
        if(l==-1)
            return false;
        aux_end=begin+l+1;
        if(*aux_end!=',')
            return false;
        a=parser(begin, aux_end);
        begin=aux_end+1;
    }
    else
        begin+=2;

    if(*(begin)!='_' && *(begin)!='(')
        return false;
    if(*(begin)=='('){
        int r=balance_brackets(begin, end);
        if(r==-1)
            return false;
        aux_end=begin+r+1;
        if(*aux_end!=')')
            return false;
        b=parser(begin, aux_end);
    }

    return a && b;
}
//PRE = { la stringa e' potenzialmente non corretta }
int binarytreebasic::balance_brackets(string::iterator i, string::iterator end){
    if(*i=='_')
        return 0;
    int k=1, s=1;
    if(end==i)//entra se la stringa è corretta e non dev'essere controllata
        end--;

    for(; k>0 && i+s!=end; s++)
        if(*(i+s)=='(')
            k++;
        else if(*(i+s)==')')
            k--;
    if(k)
        return -1;
    return s-1;
}
//POST = { ritorna il numero di passi fino alla posizione della ')' che chiude la parentesi della PRE }

binarytreebasic::nodo* binarytreebasic::somma(nodo* a, nodo* b, nodo *father){
    nodo* x=0;
    if(!a && b){
        x=new nodo(b->info, father);
        x->left=somma(a, b->left, x);
        x->right=somma(a, b->right, x);
    }
    else if(a && !b){
        x=new nodo(a->info, father);
        x->left=somma(a->left, b, x);
        x->right=somma(a->right, b, x);
    }
    else if(a && b && !a->left && !a->right && !b->left && !b->right && ((a->info->moltiplicazione(-1))==b->info))
        return x;
    else if(a && b){
        x=new nodo(((a->info)->somma(b->info)), father);
        x->left=somma(a->left, b->left, x);
        x->right=somma(a->right, b->right, x);
    }
    return x;
}

binarytreebasic::nodo* binarytreebasic::differenza(nodo* a, nodo* b, nodo *father){
    nodo* x=0;
    if(!a && b){
        x=new nodo(((b->info)->moltiplicazione(-1)), father);
        x->left=differenza(a, b->left, x);
        x->right=differenza(a, b->right, x);
    }
    else if(a && !b){
        x=new nodo(a->info, father);
        x->left=differenza(a->left, b, x);
        x->right=differenza(a->right, b, x);
    }
    else if(a && b && !a->left && !a->right && !b->left && !b->right && (a->info==b->info))
        return x;
    else if(a && b){
        x=new nodo(((a->info)->differenza(b->info)), father);
        x->left=differenza(a->left, b->left, x);
        x->right=differenza(a->right, b->right, x);
    }
    return x;
}

binarytreebasic::nodo* binarytreebasic::moltiplicazione(nodo* a, int k, nodo* father){
    nodo* x=0;
    if(a){
        if(!k)
            return new nodo(a->info->moltiplicazione(0));
        x=new nodo(((a->info)->moltiplicazione(k)), father);
        x->left=moltiplicazione(a->left, k, x);
        x->right=moltiplicazione(a->right, k, x);
    }
    return x;
}

binarytreebasic::nodo* binarytreebasic::divisione(nodo* a, int k, nodo* father){
    nodo* x=0;
    if(a){
        x=new nodo(((a->info)->divisione(k)), father);
        x->left=divisione(a->left, k, x);
        x->right=divisione(a->right, k, x);
    }
    return x;
}

vector<tipo*> binarytreebasic::nodes(nodo* n){
    vector<tipo*> a, b;
    if(!n)
        return a;
    if(!n->left && !n->right){
        a.push_back(n->info);
        return a;
    }
    vector<tipo*> c;
    c.push_back(n->info);
    a=nodes(n->left);
    b=nodes(n->right);
    a.insert(a.end(), b.begin(), b.end());
    c.insert(c.end(), a.begin(), a.end());
    return c;
}

bool binarytreebasic::controlla_percorso(string &s){
    string::iterator begin=s.begin(), end=s.end();
    for(; begin!=end; begin++)
        if(*begin!='0' && *begin!='1')
            return false;
    return true;
}

bool binarytreebasic::controlla_input(vector<tipo*>& v, string &s){
    vector<tipo*>::iterator begin=v.begin(), end=v.end();
    int size=s.size();
    for(; begin!=end; begin++)
        if(typeid(**v.begin())!=typeid(**begin))
            return false;
    return parser(s.begin(), s.end()) && (v.size()==n_nodes_stringa(s.begin())) && (size==0 || size==5 || size%6==1);
}

string binarytreebasic::tree_to_string(nodo * r){
    if(!r)
        return "_";
    return "(*,"+tree_to_string(r->left)+","+tree_to_string(r->right)+")";
}


binarytreebasic::nodo* binarytreebasic::copia(nodo * n, nodo* father) const{
    if(!n)
        return 0;
    nodo* x=new nodo(n->info, father);
    x->left=copia(n->left, x);
    x->right=copia(n->right, x);
    return x;
}

void binarytreebasic::distruggi(binarytreebasic::nodo* n){
    if(n){
        distruggi(n->left);
        distruggi(n->right);
        delete n;
    }
}

//PRE=(la stringa s e' corretta e l'albero contiene almeno un nodo, percio' il caso limite sara' "(*, _,_)")
binarytreebasic::nodo* binarytreebasic::constrRic(string::iterator i, vector<tipo*>::iterator& A, nodo* father){
    tipo* val=*A;
    A++;
    i+=3;
    nodo* x=new nodo(val, father);
    int left=balance_brackets(i, i);
    int right=balance_brackets(i+left+2, i+left+2);
    if(right)
        x->right=constrRic(i+left+2, A, x);
    if(left)
        x->left=constrRic(i, A, x);
    return x;
}

binarytreebasic::binarytreebasic(vector<tipo*> type, string& s) throw(input_error){
    if(controlla_input(type, s)){
        //int k=0;//indice valore da analizzare nell'array
        //constrRic costruisce l'albero prendendo in input l'array di valori,
        //la sua lunghezza, un iteratore iniziale e finale della stringa, l'indice del
        //valore gia' analizzato passasto per riferimento e il valore del noto padre(che inizialmente e' 0)
        vector<tipo*>::iterator it=type.begin();
        root=constrRic(s.begin(), it);
    }
    else
        throw input_error();
}

binarytreebasic::~binarytreebasic(){if(root) distruggi(root);}

binarytreebasic& binarytreebasic::operator=(const binarytreebasic& t){
    if(this != &t){
        distruggi(root);
        root=copia(t.root);
    }
    return *this;
}

int binarytreebasic::n_nodes(nodo* n){
    if(!n)
       return 0;
   return 1+n_nodes(n->left)+n_nodes(n->right);
}

tipo* binarytreebasic::cerca(string s)const throw(path_error){
    nodo* x=this->root;
    if(!controlla_percorso(s) || !x)
        throw path_error();
    string::iterator begin=s.begin(), end=s.end();
    for(; x && begin!=end; begin++)
        if(*begin=='0')
            x=x->left;
        else
            x=x->right;
    if(x)
        return x->info;
    cout<<"Elemento non trovato\n";
    return 0;
}

string binarytreebasic::struttura_tree()const{
    return tree_to_string(this->root);
}

//PRE = { la stringa e' corretta e begin punta a '(' }
unsigned int binarytreebasic::n_nodes_stringa(string::iterator begin){
    int count=0, id=1;
    if(*begin=='_')
        return count;
    for(int i=1; id; i++){
        if(*(begin+i)=='*')
            count++;
        else if(*(begin+i)=='(')
            id++;
        else if(*(begin+i)==')')
            id--;
    }
    //dato che la stringa è corretta il for finira'
    return count;
}

tipo* binarytreebasic::tipo_tree()const{
    if(!root)
        return 0;
    return root->info;
}

bool binarytreebasic::controlla_stringa(string s){
    int size=s.size();
    return parser(s.begin(), s.end()) && (size==0 || size==5 || size%6==1);
}

ostream& operator <<(ostream& os, const binarytreebasic& t){
    if(t.root)
        os<<binarytreebasic::nodo::stampa(t.root);
    return os;
}

string to_string(const binarytreebasic &t){
    if(t.root)
        return  binarytreebasic::nodo::stampa(t.root);
    return "";
}

//throw(0) i dati inseriti in input non sono corretti
//throw(1) il percorso inserito non e' corretto
